docs: Add TERMINOLOGY.md (#8290)
getsentry/sentry-cocoa: develop-docs/TERMINOLOGY.md

# Terminology

Glossary of terms used in the Sentry Cocoa SDK, with a focus on main-thread responsiveness.

## Apple Definitions

| Term                     | Definition                                                                                                                                                                                                                                           | Source                                                                                                                      |
| ------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------- |

> | **Hang**                 | The main thread is unresponsive for 250 ms or more. Measured as the time between CFRunLoop states `.afterWaiting` and `.beforeWaiting`.                                                                                                              | [Apple: Understanding hangs in your app](https://developer.apple.com/documentation/xcode/understanding-hangs-in-your-app)   |

Reviewer: `m`: The Apple docs always talk about the main run loop and not the main thread. I think we should do that too.

> Most of Apple’s developer tools start reporting issues when the period of unresponsiveness for the main run loop exceeds 250 ms. [...] Hang reporting only measures the time on the main run loop.

```suggestion
| **Hang**                 | The main run loop is unresponsive for 250 ms or more. Measured as the time between CFRunLoop states `.afterWaiting` and `.beforeWaiting`.                                                                                                              | [Apple: Understanding hangs in your app](https://developer.apple.com/documentation/xcode/understanding-hangs-in-your-app)   |
```

> | **Hang**                 | The main thread is unresponsive for 250 ms or more. Measured as the time between CFRunLoop states `.afterWaiting` and `.beforeWaiting`.                                                                                                              | [Apple: Understanding hangs in your app](https://developer.apple.com/documentation/xcode/understanding-hangs-in-your-app)   |

Reviewer: `l`: We could add a link to the docs for before and afterWaiting https://developer.apple.com/documentation/corefoundation/cfrunloopactivity/beforewaiting cause I had to look it up.

> | **Severe hang**          | A hang lasting longer than the watchdog timeout, typically around 20 seconds on iOS. The OS terminates the app.                                                                                                                                      | [Apple: Addressing watchdog terminations](https://developer.apple.com/documentation/xcode/addressing-watchdog-terminations) |

Reviewer 2: `m`: I can't find Apple using severe hang to describe watchdog terminations - only in Instruments where they use microhang, hangs, and severe hangs (I can't find the exact timing for the latter)

Author: That is exactly the problem we are facing that Apple themselves are not consistent with naming.

Reviewer 2: Indeed 😕 In that case, I'd say we should loosen the definition here so it's not confusing

Author: Changed it to "(Also) Hang"

> | (Also) **Hang**          | A hang lasting longer than the watchdog timeout, typically around 20 seconds on iOS. The OS terminates the app.                                                                                                                                      | [Apple: Addressing watchdog terminations](https://developer.apple.com/documentation/xcode/addressing-watchdog-terminations) |

Reviewer: `m`: I guess that this is a fatal app hang as mentioned below.
```suggestion
| **Fatal Hang**          | A hang lasting longer than the watchdog timeout, typically around 20 seconds on iOS. The OS terminates the app.                                                                                                                                      | [Apple: Addressing watchdog terminations](https://developer.apple.com/documentation/xcode/addressing-watchdog-terminations) |
```

Author: This is the section about Apple's definition, not ours. I can't find Apple documentation using "fatal" in the name.

> | **Hitch**                | A frame that misses its VSYNC display deadline in the render pipeline. Measured as time past the expected frame presentation. This is not a runloop concept. Hitches are about the render pipeline, covering the commit, render, and display phases. | [Apple: Explore UI animation hitches and the render loop](https://developer.apple.com/videos/play/tech-talks/10855/)        |

Reviewer 2: `l`: Apple seems to call this a 'render hitch' rather than 'hitch', should we reflect that here?

Author: I'll update it to reflect "Commit Hitches" and "Render Hitches"

> | **Render hitch**         | A frame that misses its VSYNC display deadline in the render pipeline. Measured as time past the expected frame presentation. This is not a runloop concept. Hitches are about the render pipeline, covering the commit, render, and display phases. | [Apple: Explore UI animation hitches and the render loop](https://developer.apple.com/videos/play/tech-talks/10855/)        |

Reviewer: `m`: The description on the commit hitch below mentions that render hitches happen out of process and can't be observed. It would be great to mention this here as well.

Author: Outdated; I rewrote the definitions based on the documentation

> | **Commit hitch**         | A hitch caused by the main thread missing the commit deadline of around 8 to 16 ms depending on refresh rate. This is one subtype of hitch. The other subtypes (render and display hitches) happen out of process and cannot be observed by the SDK. | [Apple WWDC 2021: Demystify and eliminate hitches](https://developer.apple.com/videos/play/wwdc2021/10252/)                 |

Reviewer: `m`: Should we maybe add an extra point for display hitches? We have a point for render hitches above, but we don't have one for display hitches.

Author: I was wrong with that definition, there are not display hitches

> | **Commit hitch**         | A hitch caused by the main thread missing the commit deadline of around 8 to 16 ms depending on refresh rate. This is one subtype of hitch. The other subtypes (render and display hitches) happen out of process and cannot be observed by the SDK. | [Apple WWDC 2021: Demystify and eliminate hitches](https://developer.apple.com/videos/play/wwdc2021/10252/)                 |

Reviewer: `m`: So, if I'm not mistaken, the commit hitch is what we can measure by also observing the main run loop. It's basically a short app hang, if I'm not mistaken. If yes, I think we should mention this info here.

Author: I rewrote the definitions; please take another look

> | **Watchdog termination** | The OS kills an app that has been unresponsive for too long, typically around 20 seconds. Detected post-mortem by comparing app state across launches.                                                                                               | [Apple: Addressing watchdog terminations](https://developer.apple.com/documentation/xcode/addressing-watchdog-terminations) |

Reviewer: `m`: I'm curious where the 20 seconds stem from. I'm unsure if that's accurate. As far as I know, Apple doesn't specify exactly when its watchdog kills your app.

Author: The example crash report is showing it [here](https://developer.apple.com/documentation/xcode/addressing-watchdog-terminations#Interpret-the-app-responsiveness-watchdog-information) with 19.97s

```
Termination Description: SPRINGBOARD, 
    scene-create watchdog transgression: application<com.example.MyCoolApp>:667
    exhausted real (wall clock) time allowance of 19.97 seconds 
```

Reviewer: Okay thanks it's clearer now.

## Sentry Definitions

These are the terms used in the Sentry SDK and product. They are inspired by Apple definitions but adapted for Sentry error reporting.

### Hang Detection

| Term              | Definition                                                                                                                                                                                                                                                                                                                                           |
| ----------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | **Hang**          | The main thread is unresponsive, detected by measuring the time the runloop is blocked between `.afterWaiting` and `.beforeWaiting`. The duration is the total blocked time, including the per-frame budget. A 500 ms hang means the runloop was blocked for 500 ms. Reported as an issue with a sampled profile of the main thread during the hang. |

Reviewer: `m`: 
```suggestion
| **Hang**          | The main run loop is unresponsive, detected by measuring the time the main run loop is blocked between `.afterWaiting` and `.beforeWaiting`. The duration is the total blocked time, including the per-frame budget. A 500 ms hang means the runloop was blocked for 500 ms. Reported as an issue with a sampled profile of the main thread during the hang. |
```

> | **Commit hitch**         | A hitch caused by the app's main thread [missing the commit deadline](https://developer.apple.com/documentation/xcode/understanding-hitches-in-your-app#Understand-commit-hitches-and-render-hitches) (~[8.3 to 16.7 ms](https://developer.apple.com/videos/play/tech-talks/10855/?time=443) depending on refresh rate), preventing the Core Animation commit from reaching the render server in time. Commit hitches and hangs [share the same root cause](https://developer.apple.com/documentation/xcode/understanding-hitches-in-your-app#Overview) — an unresponsive main thread — but differ by interaction type: during continuous interactions (scrolling, animation) it's a commit hitch; during discrete interactions (tap, press) it's a hang. This is the only subtype of hitch observable by the SDK.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

Reviewer: `l`:

> In the Commit Hitch illustration above, the app’s process doesn’t meet the commit deadline, which causes the render server to start one vsync later, resulting in a hitch. - https://developer.apple.com/documentation/xcode/understanding-hitches-in-your-app#Understand-commit-hitches-and-render-hitches
```suggestion
| **Commit hitch**         | A hitch caused by the app's process [missing the commit deadline](https://developer.apple.com/documentation/xcode/understanding-hitches-in-your-app#Understand-commit-hitches-and-render-hitches) (~[8.3 to 16.7 ms](https://developer.apple.com/videos/play/tech-talks/10855/?time=443) depending on refresh rate), preventing the Core Animation commit from reaching the render server in time. Commit hitches and hangs [share the same root cause](https://developer.apple.com/documentation/xcode/understanding-hitches-in-your-app#Overview) — an unresponsive main thread — but differ by interaction type: during continuous interactions (scrolling, animation) it's a commit hitch; during discrete interactions (tap, press) it's a hang. This is the only subtype of hitch observable by the SDK.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
```

### Hitch Detection

| Term            | Definition                                                                                                                                                                                                                                                                                                                                                                                                                                |
| --------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | **Hitch**       | A runloop iteration that exceeds the frame deadline (~16 ms at 60 fps), measured only during interactive mode (touch input, scrolling). Unlike hangs, the duration is only the overshoot past the per-frame budget, not the total iteration time. A 20 ms runloop iteration at 60 fps is a 3.3 ms hitch. Hitches that cross the 250 ms threshold are tracked as hangs instead. Severity tiers based on overshoot: >10 ms, 5-10 ms, <5 ms. |

Reviewer: `m`:
```suggestion
| **Hitch**       | A main run loop iteration that exceeds the frame deadline (~16 ms at 60 fps), measured only during interactive mode (touch input, scrolling). Unlike hangs, the duration is only the overshoot past the per-frame budget, not the total iteration time. A 20 ms runloop iteration at 60 fps is a 3.3 ms hitch. Hitches that cross the 250 ms threshold are tracked as hangs instead. Severity tiers based on overshoot: >10 ms, 5-10 ms, <5 ms. |
```
